### PR TITLE
feat: Add new UDA integration support

### PR DESCRIPTION
**Pull request by jdogmcsteezy**
Also started to add Python-type hints.
TOOLS-1979 (ASADM) Display license usage data per-namespace in the summary
TOOLS-1978 (ASADM) Include the time of the latest UDA entry used in the license usage calculation.
TOOLS-1977 (ASADM) Optionally include UDA data-points where the cluster was unstable when calculating license usage by using the `--agent-unstable` flag.
TOOLS-1976 (ASADM) Optionally include the UDA's raw store in the collectinfo file using the `--agent-store` flag.

### CODE REVIEW

File: lib/live_cluster/client/cluster.py [hunk 1]
```diff
@@ -107,15 +108,35 @@ def __str__(self):
 
         return retval
 
-    def get_node_displaynames(self):
+    def get_node_displaynames(self, nodes=None):
+        selected_nodes = nodes
```
Comment (jdogmcsteezy) [inline lib/live_cluster/client/cluster.py:112]: check this logic

File: lib/live_cluster/collectinfo_controller.py [hunk 2]
```diff
@@ -49,39 +50,42 @@ def _collect_local_file(self, src, dest_dir):
         except Exception as e:
             raise e
 
-    async def _collectinfo_content(self, func, parm="", alt_parms=""):
+    async def _collectinfo_capture_and_write_to_file(
+        self, filename: str, func: Callable, param: list[Any] = []
```
Comment (jdogmcsteezy) [inline lib/live_cluster/collectinfo_controller.py:54]: param should be of type list[str]

File: lib/live_cluster/collectinfo_controller.py [hunk 3]
```diff
@@ -445,6 +436,19 @@ async def _get_collectinfo_data_json(
         snp_map[cluster_name] = dump_map
         return snp_map
 
+    def _dump_in_json_file(self, complete_name, dump):
+        self.logger.info("Dumping collectinfo {}.", complete_name)
+
+        try:
+            json_dump = json.dumps(dump, indent=2, separators=(",", ":"))
+            with open(complete_name, "w") as f:
```
Comment (jdogmcsteezy) [inline lib/live_cluster/collectinfo_controller.py:444]: use dump_collectinfo_file here

File: lib/live_cluster/collectinfo_controller.py [hunk 4]
```diff
@@ -479,263 +482,336 @@ async def _dump_collectinfo_json(
...
 
-        self._dump_in_json_file(
-            as_logfile_prefix, "aslicenseusage.json", resp, ignore_errors
-        )
-        return True
+        if "raw_store" in resp:
+            filename = "aslicenseraw.store"
+            complete_filename = as_logfile_prefix + filename
+            raw_store = resp["raw_store"]
+            try:
```
Comment (jdogmcsteezy) [inline lib/live_cluster/collectinfo_controller.py:521]: reuse dump_collectinfo_file

File: lib/live_cluster/collectinfo_controller.py [hunk 5]
```diff
@@ -479,263 +482,336 @@ async def _dump_collectinfo_json(
...
-        try:
-            await self._collectinfo_content(self._write_version)
-        except Exception as e:
-            util.write_to_file(self.aslogfile, str(e))
-            sys.stdout = sys.__stdout__
+            try:
+                show_controller = ShowController()
+                for show_param in dignostic_show_params:
+                    await self._collectinfo_capture_and_write_to_file(
+                        complete_filename, show_controller, show_param.split()
```
Comment (jdogmcsteezy) [inline lib/live_cluster/collectinfo_controller.py:652]: move stdout reset to inside _collectinfo_capture

File: lib/live_cluster/collectinfo_controller.py [hunk 6]
```diff
@@ -852,8 +936,11 @@ async def _run_collectinfo(
         "                                             1.2.3.4:3232,uid,passphrase,key_path",
         "                                             [2001::1234:10],uid,pwd",
         "                                             [2001::1234:10]:3232,uid,,key_path",
-        "    --agent-host    <host>       - Host IP of the Unique-Data-Agent to collect license data usage.",
-        "    --agent-port    <int>        - Port of the Unique-Data-Agent. Default: 8080",
+        "    --agent-host    <host>       - Host IP of the Unique Data Agent to collect license data usage.",
+        "    --agent-port    <int>        - Port of the UDA. Default: 8080",
+        "    --agent-store                - Collect the raw datastore of the UDA."
+        "    --agent-unstable             - When processing UDA entries allow instances where the cluster is",
```
Comment (jdogmcsteezy) [inline lib/live_cluster/collectinfo_controller.py:942]: remove --agent-unstable from help

File: lib/utils/common.py [hunk 7]
```diff
@@ -395,6 +396,318 @@ def _round_up(value, rounding_factor):
...
+            session.get(
+                "http://"
+                + agent_host
+                + ":"
+                + str(agent_port)
+                + "/v1/entries/range/time",
+                params=entries_params,
+            ),
+            session.get(
+                "http://" + agent_host + ":" + str(agent_port) + "/v1/health",
```
Comment (jdogmcsteezy) [inline lib/utils/common.py:469]: create constant for request string

File: lib/utils/common.py [hunk 8]
```diff
@@ -404,7 +717,130 @@ def _license_data_usage_adjustment(effective_repl, master_objects, used_bytes):
...
+            time_ = entry["time"]
+
+            total_data_bytes = entry["unique_data_bytes"]
+            cluster_result.update(total_data_bytes, time_)
+
+            if "namespaces" in entry:
+                for ns, usage in entry["namespaces"].items():
+                    ns_data_bytes = usage["unique_data_bytes"]
+                    if ns not in namespaces_result:
+                        namespaces_result[ns] = AggregateLicenseUsage(
```
Comment (jdogmcsteezy) [inline lib/utils/common.py:813]: Don't allow AggregateLicenseUsage not except val at init

File: lib/utils/common.py [hunk 9]
```diff
@@ -2021,7 +2247,7 @@ def _collect_ip_link_details(cmd=""):
     return out, None
 
 
-def _collectinfo_content(func, cmd="", alt_cmds=[]):
+def _collectinfo_content(func, cmd=[], alt_cmds=[]):
```
Comment (jdogmcsteezy) [inline lib/utils/common.py:2250]: see if default arg is evaluated at definition

File: lib/utils/util.py [hunk 10]
```diff
@@ -517,7 +569,32 @@ def get_values_from_second_level_of_dict(data, re_keys, return_type=None):
...
+def get_nested_value_from_dict(
+    data: dict[str, Any],
+    keys: list[str],
+    default_value: None = None,
+    return_type: Type[ReturnType] = str,
+) -> Union[ReturnType, None]:
+    pass
+
+
+@overload
```
Comment (jdogmcsteezy) [inline lib/utils/util.py:582]: remove overloads

File: lib/view/templates.py [hunk 11]
```diff
@@ -864,108 +1147,72 @@ def sindex_state_converter(edata):
...
+                Field("Per-Node", Projectors.Number("ns_stats", "devices_per_node")),
             ),
         ),
         Subgroup(
             "Memory",
             (
                 Field(
                     "Total",
-                    Projectors.Number("ns_stats", "memory_total"),
+                    Projectors.Func(
```
Comment (jdogmcsteezy) [inline lib/view/templates.py:1159]: could these use create_summary_total

**Review by dwelch-spike — APPROVED** [review]
Reviewed with Jesse
LGTM

**Pull request merged**